### ISSUE TITLE
About Auxiliary loss optimization

### ISSUE THREAD

**Issue by AaronXueNF** [InterDigitalInc/CompressAI#167]
Hi,     

I am confused about the 'Auxiliary loss'. It should be minimized but I don't know it's influence on the whole model's R_D performance.    

Should I adjust lr for aux optimizer according to some metrics when training my own model?

In tutorial "[Custom model](https://interdigitalinc.github.io/CompressAI/tutorials/tutorial_custom.html#)", the example looks like the Factorized Prior model, bmshj2018-factorized. The Rate and Distortion are included in main loss, and I don't know how aux loss works.

Thanks for answers.

**Comment by YodaEmbedding** [comment]
`aux_loss` adjusts **only** the `EntropyBottleneck.quantiles` to ensure that the trained distribution fits a few `target` boundary conditions. Nothing further. It also does not take any image data as input.

https://github.com/InterDigitalInc/CompressAI/blob/216af74fa09a141eaab68dd0ff2edd89629f62dc/compressai/entropy_models/entropy_models.py#L431-L434

https://github.com/InterDigitalInc/CompressAI/blob/216af74fa09a141eaab68dd0ff2edd89629f62dc/compressai/entropy_models/entropy_models.py#L378-L383

https://github.com/InterDigitalInc/CompressAI/blob/216af74fa09a141eaab68dd0ff2edd89629f62dc/compressai/optimizers/net_aux.py#L46-L57

Thus, I would say that minimizing `aux_loss` as much as possible isn't critical since all it does is make sure the support of the encoding distribution is finite, ensuring that we don't use too many symbols at runtime.

**Comment by Freed-Wu** [comment]
> it has nearly no effect on the trained RD performance.

It means user shouldn't adjust lr for aux optimizer according to some metrics when training model, really?

**Comment by YodaEmbedding** [comment]
The only time the `.quantiles` parameters are referenced during training is in `_get_medians`, which references the *.quantiles[:, :, 1]* (midpoints):

https://github.com/InterDigitalInc/CompressAI/blob/v1.2.4/compressai/entropy_models/entropy_models.py#L385-L387

...which is only used to offset prior to adding noise:

https://github.com/InterDigitalInc/CompressAI/blob/v1.2.4/compressai/entropy_models/entropy_models.py#L496-L498

The model should not be too heavily influenced by a near-constant "midpoint" bias.

---

Funnily enough:

 - One can minimize the aux loss at any point they desire. You can possibly even do it just before a training epoch. We do it continuously during training, under assume that the "midpoints" remain reasonably stable, though I wonder how true this assumption is, particularly in the first 1-3 epochs... And I wonder if it would be a better usage of the GPU to reduce the amount of `aux_optimizer.step()`s.
 - Differential optimization of aux loss is not necessary. What it does is so simple that a simple binary search on a pre-computed CDF could probably do the trick, and much, much quicker...
